### DISCUSSION TITLE
Parent macros get overriden by descendants

### DISCUSSION THREAD

**Discussion by StyleShit** [spatie/macroable#40] [Ideas]
The current mechanism overrides the parent's macros when using the same name. Check this test for example:

```PHP
Class A {
    use Macroable;
}

class B extends A {}

it('should not override parent macros', function () {
    A::macro('test', function () {
        return 'A';
    });

    B::macro('test', function () {
        return 'B';
    });

    $this->assertEquals('A', A::test());
    $this->assertEquals('B', B::test());
});
```

The first assertion will fail. That's actually a side-effect of a bigger issue - all of the class descendants will share the same macros object instead of having their own object and just inherit from the parent when needed (like pure OOP).

I'm not sure how we can solve this. The only way that came to my mind is to change the macros object structure a little bit, maybe something like this:

```PHP
protected $macros = [
    'Namespace\To\Class\A' => [
        'testMacro' => function () {
            return 'A';
        },
        'shouldBeInherited' => function () {
            return 'A';
        },
    ],
    'Namespace\To\Class\B' => [
        'testMacro' => function () {
            return 'B';
        },
    ],
];
```

And then when calling the macro, maybe use [ReflectionClass::getParentClass()](https://www.php.net/manual/en/reflectionclass.getparentclass.php) recursively, but it feels kinda complicated. WDYT?

**Comment by freekmurze** [comment]
Good remark. This package just mimics Laravel's `Macroable` trait. Maybe this discussion should be added there?

**Reply by StyleShit** [reply]
Hmm, that's a good idea. 
Should I copy-paste it, or just add a reference to this discussion?
